system_zero/src/alu/multiplication.rs

```diff
@@ -6,19 +6,50 @@ use plonky2::iop::ext_target::ExtensionTarget;
 use plonky2::plonk::circuit_builder::CircuitBuilder;
 use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
 
+use crate::alu::canonical::{
+    compute_canonical_inv, eval_u16s_canonical_check, eval_u16s_canonical_check_circuit,
+};
 use crate::registers::alu::*;
 use crate::registers::NUM_COLUMNS;
 
 pub(crate) fn generate_multiplication<F: PrimeField64>(values: &mut [F; NUM_COLUMNS]) {
-    // TODO
+    let in_1 = values[COL_MUL_ADD_INPUT_1].to_canonical_u64();
+    let in_2 = values[COL_MUL_ADD_INPUT_1].to_canonical_u64();
+    let in_3 = values[COL_MUL_ADD_INPUT_1].to_canonical_u64();
+    let output = in_1 * in_2 + in_3;
+
+    values[COL_MUL_ADD_OUTPUT_1] = F::from_canonical_u16(output as u16);
+    values[COL_MUL_ADD_OUTPUT_2] = F::from_canonical_u16((output >> 16) as u16);
+    values[COL_MUL_ADD_OUTPUT_3] = F::from_canonical_u16((output >> 32) as u16);
+    values[COL_MUL_ADD_OUTPUT_4] = F::from_canonical_u16((output >> 48) as u16);
+
+    values[COL_MUL_ADD_RESULT_CANONICAL_INV] = compute_canonical_inv(output);
```

Comment:
A comment explaining the return value of `compute_canonical_inv` might be useful here